crates/storage/nippy-jar/src/compression/zstd.rs

```diff
@@ -0,0 +1,175 @@
+use crate::{compression::Compression, NippyJarError};
+use serde::{Deserialize, Serialize};
+use std::{
+    fs::File,
+    io::{Read, Write},
+};
+use zstd::bulk::{Compressor, Decompressor};
+
+#[derive(Debug, Default, PartialEq, Serialize, Deserialize)]
+pub enum ZstdState {
+    #[default]
+    PendingDictionary,
+    Ready,
+}
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+/// Zstd compression structure. Supports a compression dictionary per column.
+pub struct Zstd {
+    /// State. Should be ready before compressing.
+    pub(crate) state: ZstdState,
+    /// Uses custom dictionaries to compress data.
+    pub(crate) use_dict: bool,
+    /// Max size of a dictionary
+    pub(crate) max_dict_size: usize,
+    /// List of column dictionaries.
+    pub(crate) raw_dictionaries: Option<Vec<Vec<u8>>>,
+    /// Number of columns to compress.
+    columns: usize,
+}
+
+impl Zstd {
+    /// Creates new [`Zstd`].
+    pub fn new(use_dict: bool, max_dict_size: usize, columns: usize) -> Self {
+        // TODO add level
+        Self {
+            state: if use_dict { ZstdState::PendingDictionary } else { ZstdState::Ready },
+            use_dict,
+            max_dict_size,
+            raw_dictionaries: None,
+            columns,
+        }
+    }
+
+    /// If using dictionaries, creates a list of [`Decompressor`].
+    pub fn generate_decompressors(&self) -> Result<Vec<Decompressor<'_>>, NippyJarError> {
+        if let Some(dictionaries) = &self.raw_dictionaries {
+            return Ok((0..self.columns)
+                .map(|column| Decompressor::with_dictionary(&dictionaries[column]))
+                .collect::<Result<Vec<_>, _>>()?)
+        }
+        Ok(vec![])
+    }
+
+    /// If using dictionaries, creates a list of [`Compressor`].
+    pub fn generate_compressors(&self) -> Result<Option<Vec<Compressor>>, NippyJarError> {
+        match self.state {
+            ZstdState::PendingDictionary => Err(NippyJarError::CompressorNotReady),
+            ZstdState::Ready => {
+                if !self.use_dict {
+                    return Ok(None)
+                }
+
+                let mut compressors = None;
+                if let Some(dictionaries) = &self.raw_dictionaries {
+                    let mut cmp = Vec::with_capacity(dictionaries.len());
+
+                    for dict in dictionaries {
+                        cmp.push(Compressor::with_dictionary(0, dict)?);
+                    }
+                    compressors = Some(cmp)
+                }
+                Ok(compressors)
+            }
+        }
+    }
+
+    /// Compresses a value using a dictionary.
+    pub fn compress_with_dictionary(
+        &self,
+        value: &[u8],
+        tmp_buf: &mut Vec<u8>,
+        handle: &mut File,
+        compressor: Option<&mut Compressor>,
```

Comment:
when it can be `None`?

Reply:
in the future (before we merge this PR), some columns won't have any kind of dictionary compression. i should be able to choose if column 1 is compressed and column 2 is not

Reply:
ah, got it, I got confused by the name `compress_with_dictionary` and assumed we always compress the value